### DISCUSSION TITLE
serialize from go code to pkl template

### DISCUSSION THREAD

**Discussion by contini-aron** [apple/pkl#230] [Ideas]
As of my understanding the go integration is perfectly capable of generating go code from a pkl template, however the serialization of go struct into a pkl template isnt a thing

that in my opinion would be really useful, ill provide an example on how and why

im currently writing a k8 operator and the custom CR (a yaml file) is defined by go structs with tags
e.g. of a single struct
```
type Deployment struct {
  ReplicaCount int16 `json:"replicaCount"`
}
```
and the pkl that would generate similar go code is 
```
class Deployment {
  replicaCount: Int16
}
```
this would allow to use the pkl template to generate the yaml for the CR while having the amazing benefits of pkl lang (type checking, custom error sintaxes... )

why not using the pkl to go code generation in this case? because the struct might need as in this case the tags for json serialization.

**Comment by holzensp** [comment]
Given a Pkl -> Go generator, "inverting" it to go Go -> Pkl isn't any easier than starting from scratch. You'd need to parse the go structs reliably. That said, if you're not running the Go code generation, _because you need extra StructTags_, why don't you use [`StructTags` as documented here](https://github.com/apple/pkl-go/blob/d1e95f843f5d2765da18a7a5bab91931e9ada62b/docs/modules/ROOT/pages/codegen.adoc#struct-tags)?

**Reply by contini-aron** [reply]
thank you, it was a feature i think i missed while reading the docs.

**Reply by holzensp** [reply]
Does that mean your issue is solved?

**Reply by contini-aron** [reply]
it does not entirely alone solve the problem to adapt the code it generates to the operator sdk types for cr generation but it is doable with some adaptation.

however i would still advocate for the feature of serialization it would be a nice feature :)
not by any means a high prio feat but a nice to have one

**Reply by bioball** [reply]
In lieu of this, you can maybe try going Go > JSON Schema > Pkl

Go to JSON Schema: https://github.com/invopop/jsonschema
JSON Schema to Pkl: https://pkl-lang.org/package-docs/pkg.pkl-lang.org/pkl-pantry/org.json_schema.contrib/current/generate/index.html

**Comment by contini-aron** [comment]
> In lieu of this, you can maybe try going Go > JSON Schema > Pkl
> 
> Go to JSON Schema: https://github.com/invopop/jsonschema JSON Schema to Pkl: https://pkl-lang.org/package-docs/pkg.pkl-lang.org/pkl-pantry/org.json_schema.contrib/current/generate/index.html

thanks ill try it :)
